### PR TITLE
Add Infinite scroll pagination to listings/index

### PR DESCRIPTION
**Pull request by gabiseabra**

### CODE REVIEW

**Comment by gusaiani** [comment]
@gabiseabra 

I'm getting this warning

```
warning.js?17bcd61:33 Warning: ListingsIndex(...): When calling super() in `ListingsIndex`, make sure to pass up the same props that your component's constructor was passed.
```

**Comment by gusaiani** [comment]
@gabiseabra My suggestion is we take care of accumulating `Marker`s in the map in this PR. Just rendering them is fine.

**Comment by rhnonose** [comment]
I'm getting the following warning:
```
Warning: Each child in an array or iterator should have a unique "key" prop.

Check the top-level render call using <styles>. See https://fb.me/react-warning-keys for more information.
    in IntersectionObserver (at index.js:38)
    in InfiniteScroll (at index.js?entry:399)
    in div (at index.js?entry:395)
    in div (at index.js?entry:359)
    in main (at index.js:24)
    in div (at index.js:22)
    in Layout (at index.js?entry:335)
    in ListingsIndex (at _document.js?entry:8)
    in StyleSheetManager
    in Unknown
    in AppContainer
    in Container
    in div
    in App
> Disposing inactive page(s): /, /listings, /listings/
```

**Comment by gabiseabra** [comment]
Both warnings are fixed.

**Comment by gusaiani** [comment]
@gabiseabra I think we can safely set the infinite scroll threshold to a much larger number, perhaps 400 or even 500px.

The idea is for the backend to load the new listings with a lot of leeway.

**Comment by gusaiani** [comment]
@gabiseabra In what situations would a user need to click in the button to load more?

I'm thinking we may simply do without it for now.

**Comment by gabiseabra** [comment]
Right now it doesn't do anything useful, but, with some error handling, it could act as a "retry" button when the page doesn't load.

**Comment by gusaiani** [comment]
@gabiseabra I see three issues with the button as it is right now.

1. If you really click on it now, it removes markers from the map
2. It won't do anything when you *really* see it: after you've scrolled all the way to the bottom. 
And this will happen very often, just filter to one neighborhood.
3. the styling needs improvement, perhaps no border, white or transparent background, $blue text.

Knowing the next feature will be ways to show more listings, I think we're safe without the button for the time being.

File: components/listings/index/Map/index.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import Map from 'components/shared/Map'
+import Marker from './Marker'
+
+const diff = (a, c) => Math.abs(a - c)
```
Comment (gusaiani) [inline components/listings/index/Map/index.js]: @gabiseabra What are `a` and `c`?

File: components/listings/index/Map/index.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+const relevance = (d, c) => Math.abs(d / c - 1)
```
Comment (gusaiani) [inline components/listings/index/Map/index.js]: @gabiseabra What are `d` and `c`?

File: components/shared/InfiniteScroll/styles.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import * as colors from 'constants/colors'
+
+export default styled.ul`
+  margin: 0;
+  padding: 0;
+  list-style: none;
+`
+
+export const Page = styled.li`
+`
```
Comment (gusaiani) [inline components/shared/InfiniteScroll/styles.js]: @gabiseabra Is this right?

File: services/listing-api.js [hunk 4]
```diff
@@ -43,8 +43,9 @@ const translatedKeys = {
 // { neighboorhood: ['Copacabana', 'Leblon'], max_price: '100000' }
 const buildGetParams = function(query) {
   return Object.keys(query).reduce(function(previous, key) {
-    previous[translatedKeys[key]] = splitParam(query[key], key)
-    return previous
+    const realKey = translatedKeys[key] || key
```
Comment (gusaiani) [inline services/listing-api.js]: @gabiseabra I'm not 100% sure if this is the cause, but the filter isn't working any longer.

File: components/listings/index/Map/index.js (hunk 1 above)
Comment (gabiseabra) [reply components/listings/index/Map/index.js]: Any two numbers, the relationship is not important

File: components/listings/index/Map/index.js (hunk 2 above)
Comment (gabiseabra) [reply components/listings/index/Map/index.js]: This one was just a joke, but I better rename the params to make it clearer.
a = position in the list relative to the current page
b = length of the list

File: services/listing-api.js (hunk 4 above)
Comment (gabiseabra) [reply services/listing-api.js]: This bit is correct, the problem was the `ListingsIndex` component wasn't passing `getListings` the filter params

File: components/shared/InfiniteScroll/index.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import Container, {Footer} from './styles'
+
+export default class InfiniteScroll extends Component {
+  static defaultProps = {
+    threshold: 2000
+  }
+
+  componentDidMount() {
+    window.addEventListener('scroll', this.onScroll)
+  }
```
Comment (gusaiani) [inline components/shared/InfiniteScroll/index.js]: @gabiseabra, what about adding `removeEventListener` to `componentWillUnmount`, do you think it helps?

File: pages/_document.js [hunk 6]
```diff
@@ -2,15 +2,19 @@ import Document, {Head, Main, NextScript} from 'next/document'
 import {ServerStyleSheet} from 'styled-components'
 import flush from 'styled-jsx/server'
 
+const FEATURES = ['Array.prototype.findIndex']
```
Comment (gusaiani) [inline pages/_document.js]: @gabiseabra Are we still using array’s `findIndex`?

If not, we would do go to delete the pollyfill link as well in 
https://github.com/em-casa/frontend/pull/147/files#diff-d8391eb435a71ebed23dc1b153195638R33

File: components/shared/InfiniteScroll/index.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+  onScroll = _.throttle((e) => {
```
Comment (gusaiani) [inline components/shared/InfiniteScroll/index.js]: @gabiseabra I would merge as is, but perhaps we can help our future selves by doing what's now inside `onLoad` like:

```
distanceFromFooterToWindowBottom = () => {
  const footerRectangle = this.footer.getBoundingClientRect()
  return Math.abs(footerRectangle.y - window.innerHeight)
}

shouldLoadMoreListings = () => {
  const {threshold} = this.props

  return distanceFromFooterToWindowBottom <= threshold
}

onScroll = _.throttle((e) => {
  const {onLoad} = this.props

  if (this.shouldLoadMoreListings) onLoad
}, 500)
```

File: components/shared/InfiniteScroll/index.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    if (!this.footer) return
+    const {threshold, onLoad} = this.props
```
Comment (gusaiani) [inline components/shared/InfiniteScroll/index.js]: @gabiseabra, also, the name `onLoad` seems non descriptive.

I may be totally wrong here, but shouldn't it be more like 'doLoad'?

Comment (gabiseabra) [reply components/shared/InfiniteScroll/index.js]: React component's event handlers are typically called onSomething.
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-handler-names.md

**Pull request merged**